### PR TITLE
Set author display name on Post / Page creation, send author changes to API.

### LINKED ISSUE #9015: Post/Page Options: Allow users to change the post/page author
Feature request brought via Twitter to add the "change author" functionality to the iOS app. 

Tweet: https://twitter.com/WegrynENT/status/979936474090082304

version 9.6

### PR DESCRIPTION
**Pull request by twstokes**
This is one PR in a series towards implementing #9015.

This PR is branched from https://github.com/wordpress-mobile/WordPress-iOS/pull/16298, so it should be merged first.

### Affected areas:

This PR aims to accomplish two things:

1. Sending an updated authorID when an author is changed for a Post or Page.
**Why:** To support a future PR (https://github.com/wordpress-mobile/WordPress-iOS/pull/16281) we want the value of the author user ID to be sent to the API. Currently it's not sent. This PR updates the `RemotePost`'s `authorID` value so that it will be picked up by WordPressKit (see PR: https://github.com/wordpress-mobile/WordPressKit-iOS/pull/381) and ultimately sent to the API. REST API [reference 1](https://developer.wordpress.com/docs/api/1.2/post/sites/%24site/posts/new/), [reference 2](https://developer.wordpress.com/docs/api/1.2/post/sites/%24site/posts/%24post_ID/), XML-RPC API [reference 1](https://codex.wordpress.org/XML-RPC_WordPress_API/Posts) (see `wp.newPost` and `wp.editPost`).

2. Setting an author user ID and display name on a newly created Post or Page.
**Why:** To support a future PR (https://github.com/wordpress-mobile/WordPress-iOS/pull/16281) we want the author display name and author user ID to be known on Post or Page creation to allow changing an author for a multi-user blog. We must know both so that the user interface can show the author's name (by default, theirs), and showing the currently selected author (which is determined by the author ID) in the author list. Currently, neither of these are set.

The ability to change an author in the UI will be provided in an upcoming PR: https://github.com/wordpress-mobile/WordPress-iOS/pull/16281

### To test:

**Local changes to the author**

- Run the unit tests `PostsTests`. Manipulating this from the app interface isn't currently possible, but will be in https://github.com/wordpress-mobile/WordPress-iOS/pull/16281.

**Setting the authorID and author name on Post and Page creation**

1. Add breakpoints on `EditPageViewController:56` and `EditPostViewController:127`.
2. Create a new Post or Page using the app.
3. When execution is paused for debugging, observe the value of the Post or Page (`newPost` or `newPage`).
4. Both the `newPost` and `newPage` should have a populated `author` and `authorID` field.

**The authorID is now sent on Post / Page creation and modification**

1. Add a breakpoint on `PostService.m:266`.
2. Create or update a Post or Page.
3. Observe that the value for `remotePost` contains the `authorID`.

## Regression Notes
1. Potential unintended areas of impact

- Creating or updating posts or pages.

2. What I did to test those areas of impact (or what existing automated tests I relied on)

- Created a new Post on a WordPress.com and self-hosted blog.
- Updated a Post on a WordPress.com and self-hosted blog.

3. What automated tests I added (or what prevented me from doing so)

- PostTests - testLocalChangesWhenAuthorIsChanged
- PostTests - testLocalChangesWhenAuthorIsTheSame

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WordPress/Classes/Services/BlogService+BlogAuthors.swift
Comment (twstokes) [inline WordPress/Classes/Services/BlogService+BlogAuthors.swift]: We remove old authors so they'll no longer appear in the author list, however, this also means their name and avatar will no longer show in the History view.

Comment (guarani) [reply WordPress/Classes/Services/BlogService+BlogAuthors.swift]: Is this an intentional change or a side-effect of fixing the bug on the post list screen. Could you please expand on the upsides/downsides to this change?

**Review by guarani — CHANGES_REQUESTED** [review]
Thanks for catching this bug where self-hosted post authors don't show up properly in the post list. Would this be best addressed in a new issue and this PR target `develop` instead of the feature branch?

I smoke tested this on a self-hosted site and it looks to fix the issue on the post list screen of authors not showing up. I had trouble testing the effects to the edit history screen, I couldn't get the History option to appear in the bottom sheet in the editor when I tap the three-dot menu. I might be missing something, did you run into this?

---

A few comments on the PR itself:

> This PR may help resolve a UI bug where a WordPress.org site doesn't show the author in the list of posts when viewing for "Everyone", and the user is an Administrator.

If this bug is independent from https://github.com/wordpress-mobile/WordPress-iOS/issues/9015, that would be a point in favor of making a separate issue for this bug.

I'm not clear from the PR whether this bug affects only self-hosted (WordPress.org) sites, or also affects WP.com (a separate issue for this bug could help us triage it).

Could the testing steps be more descriptive to specify what to expect with regard to the Post list screen?

Just a few nitpicks:
- Is there a PR title that better describes the issue?
- The screenshots on the PR are very large (I sometimes use a HTML img tag to adjust the width to something like 350px).

File: WordPress/Classes/Services/BlogService+BlogAuthors.swift
Comment (twstokes) [reply WordPress/Classes/Services/BlogService+BlogAuthors.swift]: Sure @guarani - glad to.

### Post list screen bug:

The bug where authors may not show on the post list screen is resolved [by this piece](https://github.com/wordpress-mobile/WordPress-iOS/pull/16269/files#diff-427e9e088f6cb54084a18b28f4841c71720a2afbaf9fc1b0edb66d6b956bec8aR815), which was put into place so that the settings view would properly show the author's name when possible for self-hosted sites. If you think it's a better idea to break this out into a separate bug I'll be happy to.

### For the code above to sync authors, here are the upsides and downsides:

**Upsides**
It should prevent authors who have been removed from the blog from being an option in the list when setting an author.

**Downsides**
When viewing the edit history for a Page or Post, the metadata about an author won't be available (e.g. avatar, display name). The query for that [is here](https://github.com/wordpress-mobile/WordPress-iOS/pull/16269/files#diff-2674017ddcb31e52b0d3f72b21bd6aceb1cd40ab83afb0fc8dd6a3fddd507f1eR99).

Another slightly more complex option would be to retain all authors, but add an extra field to mark them as "removed".

**Comment by twstokes** [comment]
> Thanks for catching this bug where self-hosted post authors don't show up properly in the post list. Would this be best addressed in a new issue and this PR target `develop` instead of the feature branch?

I think that's a great idea, will do.

> I smoke tested this on a self-hosted site and it looks to fix the issue on the post list screen of authors not showing up. I had trouble testing the effects to the edit history screen, I couldn't get the History option to appear in the bottom sheet in the editor when I tap the three-dot menu. I might be missing something, did you run into this?

I ran into this as well, and assumed it's not a feature of self-hosted sites. I'll look into that some more. 👍

> If this bug is independent from #9015, that would be a point in favor of making a separate issue for this bug.

Agreed. It was a "nice bonus" that appeared, but I'll make it the main driver for the fix.

> I'm not clear from the PR whether this bug affects only self-hosted (WordPress.org) sites, or also affects WP.com (a separate issue for this bug could help us triage it).
> 
> Could the testing steps be more descriptive to specify what to expect with regard to the Post list screen?

Sure - I'll make this clear in the dedicated bug issue for it.

> * Is there a PR title that better describes the issue?

Agreed, will improve that.

> * The screenshots on the PR are very large (I sometimes use a HTML img tag to adjust the width to something like 350px).

Good call, will do. Thanks!

**Comment by twstokes** [comment]
> Thanks for catching this bug where self-hosted post authors don't show up properly in the post list. Would this be best addressed in a new issue and this PR target `develop` instead of the feature branch?
> 
> I smoke tested this on a self-hosted site and it looks to fix the issue on the post list screen of authors not showing up. I had trouble testing the effects to the edit history screen, I couldn't get the History option to appear in the bottom sheet in the editor when I tap the three-dot menu. I might be missing something, did you run into this?
> 
> A few comments on the PR itself:
> 
> > This PR may help resolve a UI bug where a WordPress.org site doesn't show the author in the list of posts when viewing for "Everyone", and the user is an Administrator.
> 
> If this bug is independent from #9015, that would be a point in favor of making a separate issue for this bug.
> 
> I'm not clear from the PR whether this bug affects only self-hosted (WordPress.org) sites, or also affects WP.com (a separate issue for this bug could help us triage it).
> 
> Could the testing steps be more descriptive to specify what to expect with regard to the Post list screen?
> 
> Just a few nitpicks:
> 
> * Is there a PR title that better describes the issue?
> * The screenshots on the PR are very large (I sometimes use a HTML img tag to adjust the width to something like 350px).

The PR should be updated now to address these issues. Thanks!

**Review by guarani — COMMENTED** [review]
> This code should have no effect on the current app.

Would it be possible to have testing steps, even if they involve diving into code in the IDE (e.g. putting breakpoints on certain lines and printing values, etc.)?

Also, could you please link to the API documentation showing where these parameters are expected when publishing a post (i.e. author display name, etc)? At first glance it seems like a post wouldn't need more than an authorID, but I can imagine that it might need display name in case the author is then deleted. Just want to corroborate with the API documentation here.

**Comment by twstokes** [comment]
> > This code should have no effect on the current app.
> 
> Would it be possible to have testing steps, even if they involve diving into code in the IDE (e.g. putting breakpoints on certain lines and printing values, etc.)?
> 
> Also, could you please link to the API documentation showing where these parameters are expected when publishing a post (i.e. author display name, etc)? At first glance it seems like a post wouldn't need more than an authorID, but I can imagine that it might need display name in case the author is then deleted. Just want to corroborate with the API documentation here.

Hi @guarani, thanks for the feedback! I've updated the description of the PR to hopefully fit these suggestions, but please let me know if I need to expand any areas.

**Comment by guarani** [comment]
I bumped the milestone to 17.3 since I just realized that there's no reason (I can think of) to try to get this into 17.2 which is cut Monday (something I'd overlooked when adding a milestone earlier).

**Comment by guarani** [comment]
Thanks for the updates, @twstokes! 

I see the build tests are failing:

| Screen Shot 2021-04-16 at 14 42 30 |
| - |

Are you able to look into that? If you don't have access, or there's an error that you don't recognize, just let me know and I can help.

**Comment by twstokes** [comment]
> Thanks for the updates, @twstokes!
> 
> I see the build tests are failing:
> 
> Screen Shot 2021-04-16 at 14 42 30
> Are you able to look into that? If you don't have access, or there's an error that you don't recognize, just let me know and I can help.

I saw that @guarani, and it looks like `xcodebuild` timed out. Since the command timed out rather than throwing a compilation error, I wonder if CircleCI had a hiccup?

Do you have the ability to try a re-run? It doesn't look like I'm privy to that. Thank you!

**Comment by twstokes** [comment]
> > Thanks for the updates, @twstokes!
> > I see the build tests are failing:
> > Screen Shot 2021-04-16 at 14 42 30
> > Are you able to look into that? If you don't have access, or there's an error that you don't recognize, just let me know and I can help.
> 
> I saw that @guarani, and it looks like `xcodebuild` timed out. Since the command timed out rather than throwing a compilation error, I wonder if CircleCI had a hiccup?
> 
> Do you have the ability to try a re-run? It doesn't look like I'm privy to that. Thank you!

Actually @guarani let me resolve these merge conflicts, first. Thanks!

File: WordPress/WordPressTest/PostTests.swift
Comment (guarani) [inline WordPress/WordPressTest/PostTests.swift:534]: Nice and simple tests ❤️

**Review by guarani — APPROVED** [review]
I tested on a WP.com site and LGTM! From your comment [here](https://github.com/wordpress-mobile/WordPressKit-iOS/pull/381#issuecomment-821661522) sounds like we won't merge this until Tuesday.

**Comment by twstokes** [comment]
Excellent. Thank you @guarani!

**Comment by guarani** [comment]
Just a heads-up that the 17.2 branch is now cut: https://github.com/wordpress-mobile/WordPress-iOS/tree/release/17.2
It should be safe to get this merged to `develop` for inclusion in 17.3.

**Comment by twstokes** [comment]
Sounds good @guarani, I'll resolve the merge conflict by pointing to the appropriate WordPressKit and push the changes.

**Comment by guarani** [comment]
I noticed the changes to the Podfile were reverted in 4f50606974e004abdc71bf494c4ab6eede490b33. I think my comment in https://github.com/wordpress-mobile/WordPressKit-iOS/pull/381#pullrequestreview-638123652 might have been misleading. With the second option I mentioned there, we still need to tag a new WordPressKit beta release after merging its PR and we then need to point this PR to it, and merge this.

I've gone ahead and made the release here: https://github.com/wordpress-mobile/WordPressKit-iOS/releases

When this PR is updated to point to the tag, I can trigger the CI to run again and we can get this merged 👍

**Comment by twstokes** [comment]
Sorry I misunderstood @guarani! I believe this is now resolved with the latest push.

**Comment by guarani** [comment]
I've enabled auto-merge and re-triggered the CI, so with a bit of luck this should get automatically merged once CI has passed.

**Pull request merged**
